drivers/dfplayer/dfplayer_internal.c

```diff
@@ -0,0 +1,488 @@
+
+        /* wait for reply from ISR */
+        dev->sync = thread_getpid();
+        thread_sleep();
+        dev->sync = KERNEL_PID_UNDEF;
+
+        /* In case no timeout occurred, stop timer */
+        xtimer_remove(&timer);
+
+        if (dev->flags & DFPLAYER_FLAG_TIMEDOUT) {
+            DEBUG("[dfplayer] Response timed out\n");
+            retval = -ETIMEDOUT;
+        }
+        else {
+            uint8_t code = dev->buf[0];
+            if (code == DFPLAYER_RESPONSE_ERROR) {
+                switch (dev->buf[3]) {
+                    case DFPLAYER_ERROR_BUSY:
+                        DEBUG("[dfplayer] Error: Module is busy\n");
+                        retval = -EAGAIN;
+                        break;
+                    case DFPLAYER_ERROR_FRAME:
+                        DEBUG("[dfplayer] Error: DFPlayer received incomplete "
+                              "frame\n");
+                        retval = -EIO;
+                        break;
+                    case DFPLAYER_ERROR_FCS:
+                        DEBUG("[dfplayer] Error: DFPlayer received corrupt frame "
+                              "(FCS mismatch)\n");
+                        retval = -EIO;
+                        break;
+                    default:
+                        DEBUG("[dfplayer] Unknown error!\n");
+                        /* This should never be reached according the datasheet */
+                        retval = -EIO;
+                        break;
+                }
+            }
+        }
+
+        /* wait to work around HW bug */
+        xtimer_usleep(DFPLAYER_SEND_DELAY_MS * US_PER_MS);
+
+        if (!retval) {
+            break;
+        }
+    }
+
+    return retval;
+}
+
+int dfplayer_transceive(dfplayer_t *dev, uint16_t *resp,
+                        uint8_t cmd, uint8_t p1, uint8_t p2)
+{
+    if (!dev) {
+        return -EINVAL;
+    }
+
+    mutex_lock(&dev->mutex);
+
+    int retval = _send(dev, cmd, p1, p2,  DFPLAYER_TIMEOUT_MS * US_PER_MS);
+    if (retval) {
+        mutex_unlock(&dev->mutex);
+        return retval;
+    }
+
+    if (resp) {
+        *resp = (((uint16_t)dev->buf[2]) << 8) | (uint16_t)dev->buf[3];
+    }
+
+    mutex_unlock(&dev->mutex);
+    return 0;
+}
+
+int dfplayer_reset(dfplayer_t *dev)
+{
+    if (!dev) {
+        return -EINVAL;
+    }
+
+    mutex_lock(&dev->mutex);
+
+    int retval = _send(dev, DFPLAYER_CMD_RESET, 0, 0,
+                       DFPLAYER_TIMEOUT_MS * US_PER_MS);
+
+    if (retval) {
+        mutex_unlock(&dev->mutex);
+        return retval;
+    }
+
+    /* Await notification the DFPlayer Mini is available */
+    /* set timeout timer */
+    xtimer_t timer = {
+        .callback = _timeout_cb,
+        .arg = dev
+    };
+    xtimer_set(&timer, DFPLAYER_BOOTUP_TIME_MS * US_PER_MS);
+    dev->sync = thread_getpid();
+    thread_sleep();
+    dev->sync = KERNEL_PID_UNDEF;
+
+    /* In case no timeout occurred, stop timer */
+    xtimer_remove(&timer);
```

Comment:
I think `xtimer_mutex_lock_timeout()` would be a lot shorter :wink:

Reply:
Done